Make member variables of LowPassFilter class generic

pedroazeredo04 opened the pull request:
This PR aims to tackle #345.

In order to do so,  I followed the advice from @christophfroehlich, using [`type_traits`](https://en.cppreference.com/w/cpp/header/type_traits) to identify which type was passed to the template of the `LowPassFilter` and configure the `StorageType` accordingly, all of that in compile time.

Since there is now only one set of three variables instead of two sets, I also corrected all the other occurrences of the variables were removed.

Also, @christophfroehlich said

> as an alternative, a traits struct with an initialize() method and operator overloads with a specialization for every non-trivial msg type could work and make the implementation itself cleaner. (no overload of update() would be necessary)

I thought this was a very beautiful idea, but I had some difficulties making this actually happen, because the function `update()` has some particularities for each data type. At the end, I was basically re-implementing the `update()` function three times within the traits struct. So I thought it was simpler and more readable to leave the overload the way it is. However, I am open to suggestions!!

One last thing: I had to change the base branch to `jazzy`, because the `ros2-master` branch was not compiling in my machine. Would you like for me to target the branch `ros2-master` instead of `jazzy`?

Please let me know anything that would be nice to change

saikishor reviewed (changes requested):
@pedroazeredo04 can you fix the pre-commit?. Just install pre-commit and run `pre-commit run --all-files`

saikishor commented on control_toolbox/include/control_toolbox/low_pass_filter.hpp:
```diff
@@ -123,10 +124,17 @@ class LowPassFilter
+
+  // Define the storage type based on T
+  using StorageType = typename std::conditional<
+    std::is_same<T, geometry_msgs::msg::WrenchStamped>::value, Eigen::Matrix<double, 6, 1>,
```
```suggestion
    std::is_same<T, geometry_msgs::msg::WrenchStamped>::value, Eigen::Vector6d,
```
May be we could use this directly

christophfroehlich replied:
This might come from 
https://github.com/ros/geometry/blob/noetic-devel/eigen_conversions/include/eigen_conversions/eigen_msg.h#L100
according to the comment
`  // TODO(destogl): use wrenchMsgToEigen`
But I don't find these conversion wrappers for ROS 2, https://docs.ros.org/en/rolling/p/tf2_eigen/generated/index.html#tf2-eigen

saikishor commented on control_toolbox/include/control_toolbox/low_pass_filter.hpp:
```diff
@@ -143,12 +151,18 @@ LowPassFilter<T>::~LowPassFilter()
+  if constexpr (std::is_same<T, geometry_msgs::msg::WrenchStamped>::value)
+  {
+    // TODO(destogl): make the size parameterizable and more intelligent is using complex types
+    for (Eigen::Index i = 0; i < 6; ++i)
+    {
+      filtered_value_[i] = filtered_old_value_[i] = old_value_[i] = std::numeric_limits<double>::quiet_NaN();
+    }
```
This can be completely skipped, if we define EIGEN_INITIALIZE_MATRICES_BY_NAN before including Eigen. @christophfroehlich what do you think about it?. I don't have a strong opinion here

christophfroehlich replied:
didn't know about that, why not

christophfroehlich commented:
Thanks a lot for your contribution.

> I thought this was a very beautiful idea, but I had some difficulties making this actually happen, because the function `update()` has some particularities for each data type. At the end, I was basically re-implementing the `update()` function three times within the traits struct. So I thought it was simpler and more readable to leave the overload the way it is. However, I am open to suggestions!!

You are right about that, I made a proposal in the issue: what do you think? I still think that the algorithms in the filters would be more clean, and we easily could apply that to all the other filters and add for example Pose messages (in a future PR).

pedroazeredo04 commented:
@christophfroehlich I commited your suggestion of having a `FilterTrait` struct! But I still couldn't quite make the function `LowPassFilter::update()` have an unique implementation for the three distinct data types :( 

That is because the override of this function for the type `std::vector<double>` does some additional verifications that the other data types doesn't. For instance, this block

```C
if (filtered_value_.empty())
  {
    if (std::any_of(data_in.begin(), data_in.end(), [](double val) { return !std::isfinite(val); }))
    {
      return false;
    }
    filtered_value_ = data_in;
    filtered_old_value_ = data_in;
    old_value_ = data_in;
  }
  else
  {
    assert(
      data_in.size() == filtered_value_.size() &&
      "Internal data and the data_in doesn't hold the same size");
    assert(data_out.size() == data_in.size() && "data_in and data_out doesn't hold same size");
  }
```

is very distinct from the other overrides, so it is hard to generalize this for every data type. 

---

Also, about the other methods of the `TypeTraits` struct, I could add the `isnan()`, `isempty()` and `copy_metadata()` if you like, what do you think?

> Something like this, plus other necessary member functions like `isnan`, `isempty`, `copy_metadata` (for copying the header if it is a message type) or operator overloads for `+operator`, `*operator`, which can go in its own header file:

I would only have to say , in order to do an operator overload for `+operator` and `*operator` for every data type, I think we would have to create a wrapper struct for each non-primitive type, because we probably shouldn't overload the `+operator` of `std::vector<double>`, for example, as it is [undefined behavior](https://en.cppreference.com/w/cpp/language/extending_std). What I mean is that instead of 

```C
  // undefined behavior
  template<typename T>
  std::vector<T> operator+(const std::vector<T>& lhs, const std::vector<T>& rhs) {
    std::vector<T> result(lhs.size());
    for (size_t i = 0; i < lhs.size(); ++i) {
      result[i] = lhs[i] + rhs[i];
    }
    return result;
  }
```

We would probably have to do

```C
// Wrapper around std::vector<double> to enable element-wise + and * without touching std::
struct VecD {
  std::vector<double> v;
  VecD() = default;
  VecD(const std::vector<double>& data) : v(data) {}
  bool empty() const { return v.empty(); }
  void resize(size_t n) { v.resize(n); }
  double& operator[](size_t i) { return v[i]; }
  const double& operator[](size_t i) const { return v[i]; }
};

// element-wise add
inline VecD operator+(const VecD& a, const VecD& b) {
  VecD out(a.v.size());
  for (size_t i = 0; i < a.v.size(); ++i)
    out.v[i] = a.v[i] + b.v[i];
  return out;
}

```

christophfroehlich commented:
Thanks for your work on this. My suggestion was only a draft, I'm not really experienced in writing templated classes, so I might be running in a dead end. Please shout out if this is the case ;)

> @christophfroehlich I commited your suggestion of having a `FilterTrait` struct! But I still couldn't quite make the function `LowPassFilter::update()` have an unique implementation for the three distinct data types :(
> 
> That is because the override of this function for the type `std::vector<double>` does some additional verifications that the other data types doesn't. For instance, this block
> 
> ```c
> if (filtered_value_.empty())
>   {
>     if (std::any_of(data_in.begin(), data_in.end(), [](double val) { return !std::isfinite(val); }))
>     {
>       return false;
>     }
>     filtered_value_ = data_in;
>     filtered_old_value_ = data_in;
>     old_value_ = data_in;
>   }
>   else
>   {
>     assert(
>       data_in.size() == filtered_value_.size() &&
>       "Internal data and the data_in doesn't hold the same size");
>     assert(data_out.size() == data_in.size() && "data_in and data_out doesn't hold same size");
>   }
> ```
> 
> is very distinct from the other overrides, so it is hard to generalize this for every data type.

The if branch could be implemented with the already mentionend isempty(), isnan(), and the else branch could be solved by a `if constexpr` on std:.vector?
```c++
// helper: default false
template<typename T>
struct is_std_vector : std::false_type {};

// specialization for std::vector
template<typename U, typename Alloc>
struct is_std_vector<std::vector<U, Alloc>> : std::true_type {};
```
and use     `if constexpr (is_std_vector<T>::value) {}`

> Also, about the other methods of the `TypeTraits` struct, I could add the `isnan()`, `isempty()` and `copy_metadata()` if you like, what do you think?
> 
> > Something like this, plus other necessary member functions like `isnan`, `isempty`, `copy_metadata` (for copying the header if it is a message type) or operator overloads for `+operator`, `*operator`, which can go in its own header file:
> 
> I would only have to say , in order to do an operator overload for `+operator` and `*operator` for every data type, I think we would have to create a wrapper struct for each non-primitive type, because we probably shouldn't overload the `+operator` of `std::vector<double>`, for example, as it is [undefined behavior](https://en.cppreference.com/w/cpp/language/extending_std). What I mean is that instead of

This is a very good hint (haven't thought about that), but we aren't overloading this to `std::vector<double>`, but to struct `FilterTraits<std::vector<T>>`? Isn't this a difference w.r.t to your concern?

pedroazeredo04 commented:
Hi @christophfroehlich ! No need to thank my work on this, I'm having fun messing around with templates :)

I pushed a whip code that tries to make the update() function trully generic.

> The if branch could be implemented with the already mentionend isempty(), isnan(), and the else branch could be solved by a `if constexpr` on std:.vector?
> 
> ```c++
> // helper: default false
> template<typename T>
> struct is_std_vector : std::false_type {};
> 
> // specialization for std::vector
> template<typename U, typename Alloc>
> struct is_std_vector<std::vector<U, Alloc>> : std::true_type {};
> ```
> 
> and use `if constexpr (is_std_vector<T>::value) {}`

That is a good idea! I did something similar, but checking for the types inside of the function itself.

 > This is a very good hint (haven't thought about that), but we aren't overloading this to std::vector<double>, but to struct FilterTraits<std::vector<T>>? Isn't this a difference w.r.t to your concern?

In my understanding, the problem is that `FilterTraits<std::vector<T>>` isn't the type of the variables `filtered_value_`, `old_value_` and `filtered_old_value_`. Those variables _are_ `std::vector<double>`. So, in order for the block

```C
  filtered_value_ = old_value_ * b1_ + filtered_old_value_ * a1_;
  filtered_old_value_ = filtered_value_;
```

work for every possible datatype, I had to add in a `FilterVector` wrapper around the `std::vector`. So, for the `std::vector<double>` overload, the `StorageType` is actually of type `FilterVector`, and not `std::vector<double>`. At the end I though it turned out not so ugly ahahaha, but I'm of course open to suggestions. If you could review it and give your feedback, I would appreciate it!

christophfroehlich commented on control_toolbox/include/control_toolbox/filter_traits.hpp:
```diff
@@ -0,0 +1,198 @@
+
+  static bool is_nan(const StorageType & storage) { return std::isnan(storage); }
+
+  static bool is_infinite(const StorageType & storage) { return !std::isfinite(storage); }
```
i'd prefer to use the `is_finite` instead of `is_infinite`, as we are used to this from std

christophfroehlich commented on control_toolbox/include/control_toolbox/filter_traits.hpp:
```diff
@@ -0,0 +1,198 @@
+template <>
+struct FilterTraits<geometry_msgs::msg::WrenchStamped>
+{
+  using StorageType = Eigen::Matrix<double, 6, 1>;
```
should we use?
```suggestion
  using StorageType = Vector6d;
```

pedroazeredo04 replied:
I didn't found the `Eigen::Vector6d` type, and the compiler doesn't seem to like when I change to this type. Should I include some specific file in order to use this type?

christophfroehlich reviewed (changes requested):
I like the direction of the simplification of the filter itself. The filter_traits look a bit verbose, but the API is clean and understandable IMHO.

Only some minor comments from my side left

pedroazeredo04 commented on control_toolbox/include/control_toolbox/low_pass_filter.hpp:
```diff
@@ -143,128 +149,59 @@ LowPassFilter<T>::~LowPassFilter()
-    data_out[i] = b1_ * old_value[i] + a1_ * filtered_old_value[i];
-    filtered_old_value[i] = data_out[i];
-    if (std::isfinite(data_in[i]))
+    if constexpr (std::is_same_v<T, std::vector<double>>)
```
Do you think I should change that to generalize the type of the vector? If I understood correctly, I believe the previous code only supported `std::vector<double` also

christophfroehlich replied:
have you tried my code snippet above? I thought this could work. 
we could also make another method for "validate_input" or similar? could be useful for other message types also, like checking if the frame_id the same or whatever.

pedroazeredo04 commented:
Thanks for the review @christophfroehlich !!

I commited some the small fixes that you suggested, and I will try to fix the failing tests later.

pedroazeredo04 commented:
Hi @christophfroehlich. I had some time today and I added the generic method `Traits::validate_input` to the `FilterTraits` struct, as you suggested. I also found a bug in the `Eigen::Matrix` initialization, and fixed that too. Hopefully, with those changes, all tests will pass😄. If you could take a look at the code and give me your feedback when you have a chance, that would be nice!

christophfroehlich commented on control_toolbox/include/control_toolbox/filter_traits.hpp:
```diff
@@ -0,0 +1,254 @@
+
+// specialization for std::vector
+template <typename U, typename Alloc>
+struct is_std_vector<std::vector<U, Alloc>> : std::true_type
```
is this used now?

christophfroehlich commented on control_toolbox/include/control_toolbox/filter_traits.hpp:
```diff
@@ -0,0 +1,254 @@
+};
+
+template <>
+struct FilterTraits<std::vector<double>>
```
can this be generalized for all types inside a std::vector?

christophfroehlich reviewed (approved):
LGTM, 2 minor comments

saikishor commented on control_toolbox/include/control_toolbox/filter_traits.hpp:
```diff
@@ -0,0 +1,246 @@
+
+  static void assign(StorageType & storage, const StorageType & data_in) { storage = data_in; }
+
+  static void validate_input(const T & data_in, const StorageType & filtered_value, T & data_out)
+  {
+    (void)data_in;
+    (void)filtered_value;
+    (void)data_out;  // Suppress unused warnings
+  }
+
+  static void add_metadata(StorageType & storage, const StorageType & data_in)
+  {
+    (void)storage;
+    (void)data_in;
+  }
```
Can't we make it pure virtual?

pedroazeredo04 replied:
I think we can't do virtual templates (I say this based on [this discussion](https://stackoverflow.com/questions/2354210/can-a-class-member-function-template-be-virtual) here). But the discussion does mention type erasure as an alternative, which might be applicable in our case. What do you think?

saikishor commented on control_toolbox/include/control_toolbox/filter_traits.hpp:
```diff
@@ -0,0 +1,246 @@
+  }
+};
+
+template <typename U, typename Alloc>
+struct FilterTraits<std::vector<U, Alloc>>
```
Are we specialising with allocator here?

pedroazeredo04 replied:
If I understand it correctly this function will only be used by "default" vectors, so I believe we could just drop the `Alloc`, what do you think?

saikishor replied:
Yes! You are right. That's why I asked about the alloc template argument

saikishor commented on control_toolbox/include/control_toolbox/filter_traits.hpp:
```diff
@@ -0,0 +1,246 @@
+// the std::vector<U> StorageType specialization.
+// This is a workaround for the fact that
+// std::vector<U>'s operator* and operator+ cannot be overloaded.
+template <typename U>
+struct FilterVector
+{
+  std::vector<U> data;
+
+  FilterVector() = default;
+
+  explicit FilterVector(const std::vector<U> & vec) : data(vec) {}
+
+  explicit FilterVector(size_t size, const U & initial_value = U{}) : data(size, initial_value) {}
+
+  FilterVector operator*(const U & scalar) const
+  {
+    FilterVector result = *this;
+    for (auto & val : result.data)
+    {
+      val *= scalar;
+    }
+    return result;
+  }
+
+  FilterVector operator+(const FilterVector & other) const
+  {
+    assert(data.size() == other.data.size() && "Vectors must be of the same size for addition");
+    FilterVector result = *this;
+    for (size_t i = 0; i < data.size(); ++i)
+    {
+      result.data[i] += other.data[i];
+    }
+    return result;
+  }
+
+  size_t size() const { return data.size(); }
+};
```
Are we using this somewhere?

pedroazeredo04 replied:
This is being used as the [`StorageType`](https://github.com/ros-controls/control_toolbox/blob/56c7edbff44b413cb449cb9ef4ded660bcf28429/control_toolbox/include/control_toolbox/filter_traits.hpp#L185) of the `std::vector` overload of the `FilterTraits`.

The reason for that is that we want to override `operator+` and `operator*` for every DataType, so that the implementation inside [`low_pass_filter.hpp`](https://github.com/ros-controls/control_toolbox/blob/56c7edbff44b413cb449cb9ef4ded660bcf28429/control_toolbox/include/control_toolbox/low_pass_filter.hpp#L184-L185) stays the same:

```C
  // Filter
  filtered_value_ = old_value_ * b1_ + filtered_old_value_ * a1_;
  filtered_old_value_ = filtered_value_;
```

However, we cannot override `operator+` and `operator*` of `std::vector`, because that leads to undefined behavior. So I wrote a wrapper around the std::vector just so that we could override these operators.

An alternative for that would be to write an `add()` and `multiply()` method for every overload of the `FilterTraits`. But I think that would be not so nice, as the above snippet for the filter would not be so legible, because we would have to do something like: 

```C
filtered_value_ = add(multiply(old_value_, b1_), multiply(filtered_old_value_, a1_));
```

So that is why that wrapper is there.

saikishor replied:
Got it

christophfroehlich commented:
This happens if you push a new commit after a review, no worries

saikishor commented on control_toolbox/include/control_toolbox/filter_traits.hpp:
```diff
@@ -0,0 +1,246 @@
+
+  static void initialize(StorageType & storage)
+  {
+    storage.data = std::vector<U>{std::numeric_limits<U>::quiet_NaN()};
```
I'm thinking about this one. I'm not sure if we have to initialize it or not, because in the first loop it will have to fill in the data right?. Do we want to have it pre-filled with the data?

I'm just saying because we are not sure about the size at the initialization time

pedroazeredo04 replied:
That's true, I think it would be better to leave it empty.

In the **[`low_pass_filter.hpp`](https://github.com/pedroazeredo04/control_toolbox/blob/f68ad4a3f489a71646a42fa44023347783e8e8da/control_toolbox/include/control_toolbox/low_pass_filter.hpp#L166),** we already check for emptiness to check if it is a first call. So it would make sense to leave the `std::vector` empty, because, as you said, we don't know its size at the initialization time.

But after your question I thought: does it even make sense to have a [`configure()`](https://github.com/pedroazeredo04/control_toolbox/blob/f68ad4a3f489a71646a42fa44023347783e8e8da/control_toolbox/include/control_toolbox/low_pass_filter.hpp#L147-L155) method to make the filter variables be NaN? I think it would be simpler if we removed the `configure()` method for all overloads, not only the `std::vector`.

And we could check if it is the first time that the `update()` method is being called by a member `bool first_run_` or something like that. So instead of having:

```C
bool LowPassFilter<T>::update(const T & data_in, T & data_out)
{
  if (!configured_)
  {
    throw std::runtime_error("Filter is not configured");
  }
  // If this is the first call to update initialize the filter at the current state
  // so that we dont apply an impulse to the data.
  if (Traits::is_nan(filtered_value_) || Traits::is_empty(filtered_value_))
  {
  // rest of the code
  }
}
```

We could have:

```C
bool LowPassFilter<T>::update(const T & data_in, T & data_out)
{
  if (!configured_)
  {
    throw std::runtime_error("Filter is not configured");
  }
  // If this is the first call to update initialize the filter at the current state
  // so that we dont apply an impulse to the data.
  if (first_run_)
  {
   // rest of the code
  first_run_ = false
  }
}

```

And remove the `configure()` method entirely. What do you think?

saikishor replied:
```suggestion
    (void)storage;
```

pedroazeredo04 replied:
It is done!

saikishor replied:
LGTM. let me run the CI

saikishor reviewed (approved):
LGTM. Just a minor comment

christophfroehlich reviewed (approved):
Thanks again, LGTM!
